diff --ruN a/stablehlo/stablehlo/dialect/StablehloOps.td b/stablehlo/stablehlo/dialect/StablehloOps.td
--- stablehlo/stablehlo/dialect/StablehloOps.td
+++ stablehlo/stablehlo/dialect/StablehloOps.td
@@ -2388,6 +2388,10 @@
   );
   let results = (outs Variadic<HLO_Tensor>);
   let hasVerifier = 1;
+  let extraClassDeclaration = [{
+    ::mlir::Value getOperand() { return (*this)->getOperand(0); }
+    ::mlir::Type getType() { return (*this)->getResult(0).getType(); }
+  }];
 }
 
 def StableHLO_CollectiveReduceOp : StableHLO_Op<"collective_reduce",
diff --ruN a/stablehlo/stablehlo/dialect/TypeInference.cpp b/stablehlo/stablehlo/dialect/TypeInference.cpp
--- stablehlo/stablehlo/dialect/TypeInference.cpp
+++ stablehlo/stablehlo/dialect/TypeInference.cpp
@@ -3949,6 +3949,32 @@
 }
 
 LogicalResult verifyCollectiveBroadcastOp(std::optional<Location> location,
+                                          Attribute replicaGroups) {
+  auto denseGroups = dyn_cast<DenseIntElementsAttr>(replicaGroups);
+  if (denseGroups) {
+    auto replicaGroupType = cast<RankedTensorType>(denseGroups.getType());
+    if (replicaGroupType.getRank() != 2)
+      return emitOptionalError(
+          location, "replica groups should be a rank 2 tensor,",
+          "but instead it is of rank ", replicaGroupType.getRank());
+
+    auto replicaIds = denseGroups.getValues<int64_t>();
+    llvm::DenseSet<int64_t> replicaIdsSeen;
+    for (int64_t replicaId : replicaIds) {
+      if (replicaId < 0)
+        return emitOptionalError(
+            location, "replica_groups values must be positive, but was given ",
+            replicaId);
+
+      if (!replicaIdsSeen.insert(replicaId).second)
+        return emitOptionalError(location, "replica id #", replicaId,
+                                 " seen more than once");
+    }
+  }
+  return success();
+}
+
+LogicalResult verifyCollectiveBroadcastOp(std::optional<Location> location,
                                           ValueRange operands,
                                           Attribute replicaGroups,
                                           bool hasDynamicRoot) {
diff --ruN a/stablehlo/stablehlo/dialect/TypeInference.h b/stablehlo/stablehlo/dialect/TypeInference.h
--- stablehlo/stablehlo/dialect/TypeInference.h
+++ stablehlo/stablehlo/dialect/TypeInference.h
@@ -448,6 +448,9 @@
                                      Value result);
 
 LogicalResult verifyCollectiveBroadcastOp(std::optional<Location> location,
+                                          Attribute replicaGroups);
+
+LogicalResult verifyCollectiveBroadcastOp(std::optional<Location> location,
                                           ValueRange operands,
                                           Attribute replicaGroups,
                                           bool hasDynamicRoot);
diff --ruN a/stablehlo/stablehlo/tests/ops_speculatability.mlir b/stablehlo/stablehlo/tests/ops_speculatability.mlir
--- stablehlo/stablehlo/tests/ops_speculatability.mlir
+++ stablehlo/stablehlo/tests/ops_speculatability.mlir
@@ -576,23 +576,11 @@
   } : (tensor<2xf64>) -> tensor<2xf64>
   "hlo_test_speculatability.is_speculatable"(%0) : (tensor<2xf64>) -> ()
 
-  %1 = "stablehlo.collective_broadcast"(%static_arg) {
-    replica_groups = dense<[[0, 1]]> : tensor<1x2xi64>,
-    channel_handle = #stablehlo.channel_handle<handle = 0, type = 0>
-  } : (tensor<2xf64>) -> tensor<?xf64>
-  "hlo_test_speculatability.is_speculatable"(%1) : (tensor<?xf64>) -> ()
-
-  %2 = "stablehlo.collective_broadcast"(%dynamic_arg) {
-    replica_groups = dense<[[0, 1]]> : tensor<1x2xi64>,
-    channel_handle = #stablehlo.channel_handle<handle = 0, type = 0>
-  } : (tensor<?xf64>) -> tensor<2xf64>
-  "hlo_test_speculatability.is_not_speculatable"(%2) : (tensor<2xf64>) -> ()
-
-  %3 = "stablehlo.collective_broadcast"(%dynamic_arg) {
+  %1 = "stablehlo.collective_broadcast"(%dynamic_arg) {
     replica_groups = dense<[[0, 1]]> : tensor<1x2xi64>,
     channel_handle = #stablehlo.channel_handle<handle = 0, type = 0>
   } : (tensor<?xf64>) -> tensor<?xf64>
-  "hlo_test_speculatability.is_speculatable"(%3) : (tensor<?xf64>) -> ()
+  "hlo_test_speculatability.is_speculatable"(%1) : (tensor<?xf64>) -> ()
   return
 }
 

