diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
index 80bdff6..1d2327c 100644
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -4,8 +4,8 @@ load("//third_party:repo.bzl", "tf_http_archive")
 
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "8918319853fbdf9e6f6cb69e96848f913a22bc31"
-    LLVM_SHA256 = "698fcde81111e2f28c03ba3f4017bc3395246d6f30b2e05d44f6619bd9bdd0b3"
+    LLVM_COMMIT = "815edc3ff646392bfee2b381d37dd35e4b04f9c5"
+    LLVM_SHA256 = "bbc6fa4993162bdc7dd39b53927906a455940b87c782d0a37e00127f8bf8c696"
 
     tf_http_archive(
         name = name,
