diff --git a/bazel/grpc_extra_deps.bzl b/bazel/grpc_extra_deps.bzl
index c2cf0027a4..df32ef54e5 100644
--- a/bazel/grpc_extra_deps.bzl
+++ b/bazel/grpc_extra_deps.bzl
@@ -33,7 +33,7 @@ def grpc_extra_deps():
     api_dependencies()

     go_rules_dependencies()
-    go_register_toolchains()
+    go_register_toolchains(version = "1.18.4")

     apple_rules_dependencies(ignore_version_differences = ignore_version_differences)

