From 82f5075946e2569589439d500733b700a3141374 Mon Sep 17 00:00:00 2001
From: Gabriel Wu <qqbbnease1004@126.com>
Date: Fri, 6 Sep 2024 11:24:10 +0800
Subject: [PATCH] set_slice3x3 -> set_slice_3x3 (#1784)


diff --git a/include/cutlass/matrix.h b/include/cutlass/matrix.h
index ab32597e..5d8ccb3c 100644
--- a/include/cutlass/matrix.h
+++ b/include/cutlass/matrix.h
@@ -7825,7 +7825,7 @@ struct Matrix<Element_, 3, 3> {

     Matrix m;

-    m.set_slice3x3({
+    m.set_slice_3x3({
       c + x * x * one_minus_cos, x * y * one_minus_cos - z * s, x * z * one_minus_cos + y * s,
       y * x * one_minus_cos * z * s, c + y * y * one_minus_cos, y * z * one_minus_cos - x * s,
       z * x * one_minus_cos - y * s, z * y * one_minus_cos + x * s, c + z * z * one_minus_cos
@@ -7845,7 +7845,7 @@ struct Matrix<Element_, 3, 3> {

     Matrix m = Matrix::identity();

-    m.set_slice3x3({
+    m.set_slice_3x3({
       Element(1) - Element(2) * a * a, Element(-2) * a * b, Element(-2) * a * c,
       Element(-2) * a * b, Element(1) - Element(2) * b * b, Element(-2) * b * c,
       Element(-2) * a * c, Element(-2) * b * c, Element(1) - Element(2) * c * c
@@ -14005,7 +14005,7 @@ struct Matrix<Element_, 4, 4> {

     Matrix m;

-    m.set_slice3x3({
+    m.set_slice_3x3({
       c + x * x * one_minus_cos, x * y * one_minus_cos - z * s, x * z * one_minus_cos + y * s,
       y * x * one_minus_cos * z * s, c + y * y * one_minus_cos, y * z * one_minus_cos - x * s,
       z * x * one_minus_cos - y * s, z * y * one_minus_cos + x * s, c + z * z * one_minus_cos
@@ -14025,7 +14025,7 @@ struct Matrix<Element_, 4, 4> {

     Matrix m = Matrix::identity();

-    m.set_slice3x3({
+    m.set_slice_3x3({
       Element(1) - Element(2) * a * a, Element(-2) * a * b, Element(-2) * a * c,
       Element(-2) * a * b, Element(1) - Element(2) * b * b, Element(-2) * b * c,
       Element(-2) * a * c, Element(-2) * b * c, Element(1) - Element(2) * c * c
--
2.43.0

