diff --git a/BUILD.bazel b/BUILD.bazel
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -223,6 +223,7 @@ cc_library(
         "//src/google/protobuf/util:time_util",
         "//src/google/protobuf/util:type_resolver_util",
     ],
+    alwayslink = 1,
 )
 
 # This provides just the header files for use in projects that need to build
diff --git a/protobuf.bzl b/protobuf.bzl
--- a/protobuf.bzl
+++ b/protobuf.bzl
@@ -244,7 +244,7 @@ Args:
   out_type: only generated a single type of source file for languages that have
     split sources (e.g. *.h and *.cc in C++)
 """
-_proto_gen = rule(
+proto_gen = rule(
     attrs = {
         "srcs": attr.label_list(allow_files = True),
         "deps": attr.label_list(providers = [ProtoGenInfo]),
@@ -436,7 +436,7 @@ def internal_objc_proto_library(
     full_deps = [d + "_genproto" for d in deps]
 
     if proto_deps:
-        _proto_gen(
+        proto_gen(
             name = name + "_deps_genproto",
             testonly = testonly,
             srcs = proto_deps,
@@ -447,7 +447,7 @@ def internal_objc_proto_library(
 
     # Note: we need to run the protoc build twice to get separate targets for
     # the generated header and the source files.
-    _proto_gen(
+    proto_gen(
         name = name + "_genproto_hdrs",
         srcs = srcs,
         deps = full_deps,
@@ -460,7 +460,7 @@ def internal_objc_proto_library(
         tags = ["manual"],
     )
 
-    _proto_gen(
+    proto_gen(
         name = name + "_genproto",
         srcs = srcs,
         deps = full_deps,
@@ -520,7 +520,7 @@ def internal_ruby_proto_library(
 
     # Note: we need to run the protoc build twice to get separate targets for
     # the generated header and the source files.
-    _proto_gen(
+    proto_gen(
         name = name + "_genproto",
         srcs = srcs,
         deps = [s + "_genproto" for s in deps],
@@ -602,7 +602,7 @@ def internal_py_proto_library(
         # is not explicitly listed in py_libs. Instead, host system is assumed to
         # have grpc installed.
 
-    _proto_gen(
+    proto_gen(
         name = name + "_genproto",
         testonly = testonly,
         srcs = srcs,
@@ -694,7 +694,7 @@ def _source_proto_library(
     full_deps = [d + "_genproto" for d in deps]
 
     if proto_deps:
-        _proto_gen(
+        proto_gen(
             name = name + "_deps_genproto",
             testonly = testonly,
             srcs = proto_deps,
@@ -703,7 +703,7 @@ def _source_proto_library(
         )
         full_deps.append(":%s_deps_genproto" % name)
 
-    _proto_gen(
+    proto_gen(
         name = name + "_genproto",
         srcs = srcs,
         deps = full_deps,
diff --git a/src/google/protobuf/BUILD.bazel b/src/google/protobuf/BUILD.bazel
--- a/src/google/protobuf/BUILD.bazel
+++ b/src/google/protobuf/BUILD.bazel
@@ -428,6 +428,7 @@ cc_library(
         "@com_google_absl//absl/time",
         "@utf8_range//:utf8_validity",
     ],
+    alwayslink = 1,
 )
 
 cc_library(